server/connector/data_sink.cpp

```diff
@@ -92,25 +96,43 @@ void RocksDBDataSink::appendData(velox::RowVectorPtr input) {
   SDB_ASSERT(input->type()->size() == _column_ids.size(),
              "RocksDBDataSink: column oids size ", _column_ids.size(),
              " doesn't match input type size ", input->type()->size());
-  // TODO(Dronplane) implement updating PK fields
-  _row_keys.clear();
-  primary_key::Create(*input, _key_childs, _row_keys);
   SDB_ASSERT(input->type()->kind() == velox::TypeKind::ROW);
-  const auto num_columns = input->childrenSize();
-  std::string table_key = key_utils::PrepareTableKey(_object_key);
-  const auto parent_size = table_key.size();
-  velox::IndexRange all_rows(0, input->size());
-  [[maybe_unused]] const auto& input_type = input->type()->asRow();
-  for (const auto& key : _row_keys) {
-    basics::StrResize(table_key, parent_size);
-    key_utils::AppendPrimaryKey(table_key, key);
+
+  // TODO(Dronplane) implement updating PK fields
+  std::string combined_key = key_utils::PrepareTableKey(_object_key);
+  SDB_ASSERT(sizeof(ObjectId) == combined_key.size());
+  static constexpr auto kObjectKeySize = sizeof(ObjectId);
+
+  const auto num_rows = input->size();
+  _cell_keys_buffers.clear();
+  _cell_keys_buffers.reserve(num_rows);
+  for (size_t row_idx = 0; row_idx < num_rows; ++row_idx) {
+    // [object_id] [pk]
+    primary_key::Create(*input, _key_childs, row_idx, combined_key);
     VELOX_CHECK(
       _transaction
-        .GetKeyLock(&_cf, rocksdb::Slice(table_key.data(), table_key.size()),
+        .GetKeyLock(&_cf,
+                    rocksdb::Slice(combined_key.data(), combined_key.size()),
                     false, true)
         .ok(),
       "Failed to acquire row lock for table {}", _object_key.id());
+
+    auto& cell_key_buffer = _cell_keys_buffers.emplace_back();
+    // [object_id] [column_id] [pk]
+    basics::StrResize(cell_key_buffer,
+                      combined_key.size() + sizeof(catalog::Column::Id));
+    std::memcpy(cell_key_buffer.data(), combined_key.data(), kObjectKeySize);
+    std::memcpy(
```

Comment:
This logic is duplicated here and in the DeleteDataSink. Could it be moved to helper function in key_utils?